src/query/ray/ray_shared_shape.rs

```diff
@@ -0,0 +1,103 @@
+use crate::{math::Real, shape::SharedShape};
+
+use super::{Ray, RayCast, RayIntersection};
+
+
+impl RayCast for SharedShape {
+  #[inline]
+  fn cast_local_ray(&self, ray: &Ray, max_toi: Real, solid: bool) -> Option<Real> {
+      if let Some(shape) = self.as_ball() {
+        shape.cast_local_ray(ray, max_toi, solid)
+      } else if let Some(shape) = self.as_cuboid() {
+        shape.cast_local_ray(ray, max_toi, solid)
+      } else if let Some(shape) = self.as_halfspace() {
+        shape.cast_local_ray(ray, max_toi, solid)
+      } else if let Some(shape) = self.as_segment() {
+        shape.cast_local_ray(ray, max_toi, solid)
+      } else if let Some(shape) = self.as_capsule() {
+        shape.cast_local_ray(ray, max_toi, solid)
+      } else if let Some(shape) = self.as_triangle() {
+        shape.cast_local_ray(ray, max_toi, solid)
+      } else if let Some(shape) = self.as_compound() {
+        shape.cast_local_ray(ray, max_toi, solid)
+      } else if let Some(shape) = self.as_trimesh() {
+        shape.cast_local_ray(ray, max_toi, solid)
+      } else if let Some(shape) = self.as_polyline() {
+        shape.cast_local_ray(ray, max_toi, solid)
+      } else if let Some(shape) = self.as_heightfield() {
+        shape.cast_local_ray(ray, max_toi, solid)
+      } else if let Some(shape) = self.as_round_cuboid() {
+        shape.cast_local_ray(ray, max_toi, solid)
+      } else if let Some(shape) = self.as_round_triangle() {
+        shape.cast_local_ray(ray, max_toi, solid)
+      } else {
+        #[cfg(feature = "dim3")] {
+          if let Some(shape) = self.as_cylinder() {
+            shape.cast_local_ray(ray, max_toi, solid)
+          } else if let Some(shape) = self.as_cone() {
+            shape.cast_local_ray(ray, max_toi, solid)
+          } else {
+            panic!("Unsupported shape");
+          }
+        }
+        #[cfg(feature = "dim2")] {
+          if let Some(shape) = self.as_round_convex_polygon() {
+            shape.cast_local_ray(ray, max_toi, solid)
+          } else {
+            panic!("Unsupported shape");
+          }
+        }
+      }
```

Comment:
Why not just `self.0.as_ref().cast_local_ray(ray, max_toi, solid)`?

Also, you should run `cargo fmt` This project uses 4 space indentation, not 2.

Reply:
@Demindiro Ah true, much better. Updated